File: freeswitch/typescript/src/Utils/Speech/CurrencySpeller.ts
import { Noun } from "./Noun";
import { NumberNoun } from "./NumberNoun";
import { LexicalUnit } from "./LexicalUnit";
import { NumberSpeller } from "./NumberSpeller";
import { Case, NumberType, Gender, Currency } from "./Enums";

export class CurrencySpeller {
    private static folder = "sipivr.speech.currency";
    private static readonly currencyNames: { [key: string]: { major: Noun, minor: Noun } } = {};

    static initialize() {
        this.currencyNames[Currency.RUB] = { major: new Noun("Ruble", this.folder), minor: new Noun("Kopeck", this.folder).setGender(Gender.Feminine) as Noun };
        this.currencyNames[Currency.USD] = { major: new Noun("Dollar", this.folder), minor: new Noun("Cent", this.folder) };
        this.currencyNames[Currency.EUR] = { major: new Noun("Euro", this.folder), minor: new Noun("Eurocent", this.folder) };
        this.currencyNames[Currency.JPY] = { major: new Noun("JPY", this.folder).setGender(Gender.Feminine) as Noun, minor: null };
        this.currencyNames[Currency.GBP] = { major: new Noun("GBP", this.folder), minor: new Noun("Pence", this.folder) };
        this.currencyNames[Currency.CHF] = { major: new Noun("CHF", this.folder), minor: new Noun("Centime", this.folder) };
        this.currencyNames[Currency.NOK] = { major: new Noun("NOK", this.folder).setGender(Gender.Feminine) as Noun, minor: new Noun("Ore", this.folder) };
        this.currencyNames[Currency.SEK] = { major: new Noun("SEK", this.folder).setGender(Gender.Feminine) as Noun, minor: new Noun("Ore", this.folder) };
        this.currencyNames[Currency.DKK] = { major: new Noun("DKK", this.folder).setGender(Gender.Feminine) as Noun, minor: new Noun("Ore", this.folder) };
        this.currencyNames[Currency.NZD] = { major: new Noun("NZD", this.folder), minor: new Noun("Cent", this.folder) };
        this.currencyNames[Currency.AUD] = { major: new Noun("AUD", this.folder), minor: new Noun("Cent", this.folder) };
        this.currencyNames[Currency.CAD] = { major: new Noun("CAD", this.folder), minor: new Noun("Cent", this.folder) };
    }

    static spell(amount: number, currency: Currency): LexicalUnit[] {
        amount = NumberSpeller.round(amount, 2);
        let res: LexicalUnit[] = [];
        let wholePart = NumberSpeller.truncate(amount);
        let fracPart = Math.abs((amount - wholePart)) * 100;

        let cn = this.currencyNames[currency];

        if (Math.abs(amount) >= 1 || amount == 0) {
            let unit = cn.major;
            let wholeSpelling = new NumberNoun(wholePart, unit);
            NumberSpeller.decomposeNumber(wholeSpelling).forEach(f => res.push(f));
        }
        if (fracPart != 0 && cn.minor != null) {
            let unit = cn.minor;
            let fracSpelling = new NumberNoun(fracPart, unit);
            NumberSpeller.decomposeNumber(fracSpelling).forEach(f => res.push(f));
        }
        return res;
    }
}

CurrencySpeller.initialize();

File: sipivr.plugins/sipivr.currency/src/main/java/ru/sipivr/currency/core/CurrencySpeller.java
package ru.sipivr.currency.core;


import ru.sipivr.currency.enums.Currency;
import ru.sipivr.number.core.*;
import ru.sipivr.number.enums.Gender;

import java.util.HashMap;

/**
 * Created by Admin on 24.01.2016.
 */
public class CurrencySpeller {
    public static class CurrencyNames {
        private final Noun majorName;
        private final Noun minorName;

        public CurrencyNames(Noun majorName) {
            this(majorName, null);
        }

        public CurrencyNames(Noun majorName, Noun minorName) {
            this.majorName = majorName;
            this.minorName = minorName;
        }

        public Noun getMajorName() {
            return majorName;
        }

        public Noun getMinorName() {
            return minorName;
        }
    }

    private static final String folder = "sipivr.speech.currency";
    private static HashMap<Currency, CurrencyNames> currencyNames;

    static {
        currencyNames = new HashMap<>();
        currencyNames.put(Currency.RUB, new CurrencyNames(new Noun(folder, "Ruble"), (Noun) new Noun(folder, "Kopeck").setGender(Gender.Feminine)));
        currencyNames.put(Currency.USD, new CurrencyNames(new Noun(folder, "Dollar"), new Noun(folder, "Cent")));
        currencyNames.put(Currency.EUR, new CurrencyNames(new Noun(folder, "Euro"), new Noun(folder, "Eurocent")));
        currencyNames.put(Currency.JPY, new CurrencyNames((Noun) new Noun(folder, "JPY").setGender(Gender.Feminine)));
        currencyNames.put(Currency.GBP, new CurrencyNames(new Noun(folder, "GBP"), new Noun(folder, "Pence")));
        currencyNames.put(Currency.CHF, new CurrencyNames(new Noun(folder, "CHF"), new Noun(folder, "Centime")));
        currencyNames.put(Currency.NOK, new CurrencyNames((Noun) new Noun(folder, "NOK").setGender(Gender.Feminine), new Noun(folder, "Ore")));
        currencyNames.put(Currency.SEK, new CurrencyNames((Noun) new Noun(folder, "SEK").setGender(Gender.Feminine), new Noun(folder, "Ore")));
        currencyNames.put(Currency.DKK, new CurrencyNames((Noun) new Noun(folder, "DKK").setGender(Gender.Feminine), new Noun(folder, "Ore")));
        currencyNames.put(Currency.NZD, new CurrencyNames(new Noun(folder, "NZD"), new Noun(folder, "Cent")));
        currencyNames.put(Currency.AUD, new CurrencyNames(new Noun(folder, "AUD"), new Noun(folder, "Cent")));
        currencyNames.put(Currency.CAD, new CurrencyNames(new Noun(folder, "CAD"), new Noun(folder, "Cent")));
    }

    private NumberSpeller numberSpeller;

    public CurrencySpeller(NumberSpeller numberSpeller)
    {
        this.numberSpeller = numberSpeller;
    }

    public CompositeLexicalUnit spell( double amount, Currency currency )
    {
        amount = numberSpeller.round(amount, 2);
        CompositeLexicalUnit res = new CompositeLexicalUnit();
        double wholePart = numberSpeller.truncate(amount);
        double fracPart = Math.abs((amount - wholePart)) * 100;

        CurrencyNames cn = currencyNames.get(currency);

        if (Math.abs(amount) >= 1 || amount == 0)
        {
            Noun unit = cn.getMajorName();
            NumberNoun wholeSpelling = new NumberNoun(wholePart, unit);
            res.getChildren().add(numberSpeller.decomposeNumber(wholeSpelling));
        }
        if (fracPart != 0 && cn.getMinorName() != null)
        {
            Noun unit = cn.getMinorName();
            NumberNoun fracSpelling = new NumberNoun(fracPart, unit);
            res.getChildren().add(numberSpeller.decomposeNumber(fracSpelling));
        }
        return res;
    }
}
